File: src/vscripts/traps/fire_trap_cycle_a.ts
export {};

declare const _G: Record<string, any>;
declare const thisEntity: CBaseEntity;

function OnStartTouch(this: void, trigger: any): void {
    // no-op
}

function OnEndTouch(this: void, trigger: any): void {
    // no-op
}

function StartLogic(this: void, trigger: any): void {
    const triggerName = thisEntity.GetName();
    const npc = Entities.FindByNameWithin(null as any, triggerName + "_npc", thisEntity.GetAbsOrigin(), 4000) as CDOTA_BaseNPC | undefined;
    if (!npc) {
        return;
    }
    const target = Entities.FindByNameWithin(null as any, triggerName + "_target", thisEntity.GetAbsOrigin(), 4000) as CDOTA_BaseNPC | undefined;
    if (!target) {
        return;
    }

    (thisEntity as any).bindNpc = npc;
    const breatheFire = npc.FindAbilityByName("breathe_fire");
    if (!breatheFire) {
        print("ERROR: thisEntity.hBreatheFireAbility not found");
        return;
    }
    (thisEntity as any).hBreatheFireAbility = breatheFire;

    const model = triggerName + "_model";
    (thisEntity as any).modelName = model;
    (thisEntity as any).fRefireTime = 1.8;
    (thisEntity as any).fQuickRefireTime = 0.5;
    (thisEntity as any).bNextAttackIsNormal = false;
    (thisEntity as any).hTarget = target;
    (thisEntity as any).nQuickRefires = 0;
    (thisEntity as any).fNextAttackTime = GameRules.GetGameTime() + (thisEntity as any).fQuickRefireTime;

    thisEntity.SetContextThink(
        "FireTrapActivateAlternating",
        () => FireTrapActivateAlternating(),
        0
    );
}

function DisableTrap(this: void, trigger: any): void {
    (thisEntity as any).bDisabled = true;
}

function FireTrapActivateAlternating(this: void): number | undefined {
    if (!IsServer()) {
        return undefined;
    }
    if ((thisEntity as any).bDisabled) {
        return -1;
    }
    if (GameRules.IsGamePaused()) {
        return 0.5;
    }
    if (GameRules.GetGameTime() >= (thisEntity as any).fNextAttackTime) {
        if ((thisEntity as any).bNextAttackIsNormal === false) {
            return QuickRefire();
        } else {
            return NormalRefire();
        }
    }
    return 0.25;
}

function QuickRefire(this: void): number {
    DoEntFire((thisEntity as any).modelName, "SetAnimation", "bark_attack", 0.4, undefined as any, undefined as any);
    (thisEntity as any).bindNpc.SetContextThink(
        "QuickRefireCast",
        () => {
            (thisEntity as any).bindNpc.CastAbilityOnPosition(
                (thisEntity as any).hTarget.GetAbsOrigin(),
                (thisEntity as any).hBreatheFireAbility,
                -1
            );
            return undefined;
        },
        0.3
    );

    (thisEntity as any).nQuickRefires = (thisEntity as any).nQuickRefires + 1;
    if ((thisEntity as any).nQuickRefires <= 2) {
        (thisEntity as any).fNextAttackTime = GameRules.GetGameTime() + (thisEntity as any).fQuickRefireTime;
    } else {
        (thisEntity as any).bNextAttackIsNormal = true;
        (thisEntity as any).fNextAttackTime = GameRules.GetGameTime() + (thisEntity as any).fRefireTime;
        (thisEntity as any).nQuickRefires = 0;
    }
    return 0.25;
}

function NormalRefire(this: void): number {
    DoEntFire((thisEntity as any).modelName, "SetAnimation", "bark_attack", 0.4, undefined as any, undefined as any);
    (thisEntity as any).bindNpc.SetContextThink(
        "NormalRefireCast",
        () => {
            (thisEntity as any).bindNpc.CastAbilityOnPosition(
                (thisEntity as any).hTarget.GetAbsOrigin(),
                (thisEntity as any).hBreatheFireAbility,
                -1
            );
            return undefined;
        },
        0.3
    );

    (thisEntity as any).fNextAttackTime = GameRules.GetGameTime() + (thisEntity as any).fRefireTime;
    (thisEntity as any).bNextAttackIsNormal = false;
    return 0.25;
}

_G.OnStartTouch = OnStartTouch;
_G.OnEndTouch = OnEndTouch;
_G.StartLogic = StartLogic;
_G.DisableTrap = DisableTrap;
_G.FireTrapActivateAlternating = FireTrapActivateAlternating;
_G.QuickRefire = QuickRefire;
_G.NormalRefire = NormalRefire;

File: game/A1/scripts/vscripts/traps/fire_trap_cycle_a.lua
--循环形火焰陷阱
--会在单次与连续3次喷射之间交替触发

-- 陷阱在一个时间点进行初始化 如果想错开时间点  可以在锤子中设置StartLogic的delay

function OnStartTouch(trigger)

end

function OnEndTouch(trigger)
   
end




function StartLogic( trigger )
    local triggerName = thisEntity:GetName()
    -- local npc = Entities:FindByName( nil, triggerName .. "_npc" )
	local npc = Entities:FindByNameWithin(nil, triggerName .. "_npc", thisEntity:GetAbsOrigin(), 4000)
    if not npc then
        return
    end
    -- local target = Entities:FindByName( nil, triggerName .. "_target" )
	local target = Entities:FindByNameWithin(nil, triggerName .. "_target", thisEntity:GetAbsOrigin(), 4000)
    if not target then
        return
    end

    thisEntity.bindNpc = npc
	thisEntity.hBreatheFireAbility = npc:FindAbilityByName( "breathe_fire" )
	if thisEntity.hBreatheFireAbility == nil then
		print( "ERROR: thisEntity.hBreatheFireAbility not found" )
		return
	end
    local model = triggerName .. "_model"

    thisEntity.modelName = model
	thisEntity.fRefireTime = 1.8
	thisEntity.fQuickRefireTime = 0.5
	thisEntity.bNextAttackIsNormal = false
    thisEntity.hTarget = target

	thisEntity.nQuickRefires = 0
	thisEntity.fNextAttackTime = GameRules:GetGameTime() + thisEntity.fQuickRefireTime

	thisEntity:SetContextThink( "FireTrapActivateAlternating", function() return FireTrapActivateAlternating() end, 0 )
end

---------------------------------------------------------------------------

function DisableTrap( trigger )
	thisEntity.bDisabled = true
end

---------------------------------------------------------------------------

function FireTrapActivateAlternating()
	if not IsServer() then
		return
	end

	if thisEntity.bDisabled then
		return -1
	end

	if GameRules:IsGamePaused() == true then
		return 0.5
	end

	if GameRules:GetGameTime() >= thisEntity.fNextAttackTime then
		if thisEntity.bNextAttackIsNormal == false then
			return QuickRefire()
		else
			return NormalRefire()
		end
	end

	return 0.25
end

---------------------------------------------------------------------------

function QuickRefire()

    DoEntFire(  thisEntity.modelName, "SetAnimation", "bark_attack", .4, self, self )
	
    thisEntity:GameTimer(0.3, function()
        thisEntity.bindNpc:CastAbilityOnPosition(  thisEntity.hTarget:GetAbsOrigin(), thisEntity.hBreatheFireAbility, -1 )
    end)
	thisEntity.nQuickRefires = thisEntity.nQuickRefires + 1

	if thisEntity.nQuickRefires <= 2 then
		thisEntity.fNextAttackTime = GameRules:GetGameTime() + thisEntity.fQuickRefireTime
	else
		thisEntity.bNextAttackIsNormal = true
		thisEntity.fNextAttackTime = GameRules:GetGameTime() + thisEntity.fRefireTime
		thisEntity.nQuickRefires = 0 -- reset counter
	end

	return 0.25
end

---------------------------------------------------------------------------

function NormalRefire()

    DoEntFire(  thisEntity.modelName, "SetAnimation", "bark_attack", .4, self, self )
	-- thisEntity:CastAbilityOnPosition( thisEntity:GetTrapTarget(), thisEntity.hBreatheFireAbility, -1 )
    thisEntity:GameTimer(0.3, function()
        thisEntity.bindNpc:CastAbilityOnPosition( thisEntity.hTarget:GetAbsOrigin(), thisEntity.hBreatheFireAbility, -1 )
    end)

	thisEntity.fNextAttackTime = GameRules:GetGameTime() + thisEntity.fRefireTime
	thisEntity.bNextAttackIsNormal = false

	return 0.25
end

---------------------------------------------------------------------------

